accord-core/src/main/java/accord/local/Command.java

```diff
@@ -19,362 +19,487 @@
-            // unlike in the Accord paper, we partition shards within a node, so that to ensure a total order we must either:
-            //  - use a global logical clock to issue new timestamps; or
-            //  - assign each shard _and_ process a unique id, and use both as components of the timestamp
-            Timestamp witnessed = txnId().compareTo(max) > 0 && txnId().epoch >= commandStore().latestEpoch()
-                    ? txnId() : commandStore().uniqueNow(max);
-            executeAt(witnessed);
+    public AcceptOutcome preaccept(SafeCommandStore safeStore, PartialTxn partialTxn, AbstractRoute route, @Nullable RoutingKey progressKey)
+    {
+        if (promised().compareTo(Ballot.ZERO) > 0)
+            return AcceptOutcome.RejectedBallot;
 
-            txn.keys().foldl(commandStore().ranges().since(txnId().epoch), (i, key, param) -> {
-                if (commandStore().hashIntersects(key))
-                    commandStore().commandsForKey(key).register(this);
-                return null;
-            }, null);
-        }
+        return preacceptInternal(safeStore, partialTxn, route, progressKey);
     }
 
-    public boolean preaccept(Txn txn, Key homeKey, Key progressKey)
+    private AcceptOutcome preacceptInternal(SafeCommandStore safeStore, PartialTxn partialTxn, AbstractRoute route, @Nullable RoutingKey progressKey)
     {
-        if (promised().compareTo(Ballot.ZERO) > 0)
+        if (executeAt() != null)
         {
-            logger.trace("{}: skipping preaccept - witnessed higher ballot ({})", txnId(), promised());
-            return false;
+            logger.trace("{}: skipping preaccept - already preaccepted ({})", txnId(), status());
+            return AcceptOutcome.Redundant;
         }
 
-        if (hasBeen(PreAccepted))
+        Status status = status();
+        switch (status)
         {
-            logger.trace("{}: skipping preaccept - already preaccepted ({})", txnId(), status());
-            return true;
+            default: throw new IllegalStateException();
+            case NotWitnessed:
+            case AcceptedInvalidate:
+            case Invalidated:
         }
 
-        witness(txn, homeKey, progressKey);
-        boolean isProgressShard = progressKey != null && handles(txnId().epoch, progressKey);
-        commandStore().progressLog().preaccept(this, isProgressShard, isProgressShard && progressKey.equals(homeKey));
+        KeyRanges coordinateRanges = coordinateRanges(safeStore);
+        ProgressShard shard = progressShard(safeStore, route, progressKey, coordinateRanges);
+        if (!validate(KeyRanges.EMPTY, coordinateRanges, shard, route, Set, partialTxn, Set, null, Ignore))
+            throw new IllegalStateException();
 
-        logger.trace("{}: preaccepted with executeAt: {}, deps: {}", txnId(), executeAt(), savedDeps());
-        notifyListeners();
-        return true;
+        Timestamp max = safeStore.maxConflict(partialTxn.keys());
```

Comment:
We talked about renaming the variable to maxConflictingTxn

Reply:
Looking at it again, I'm not 100% sure this is an improvement, as `Txn` suffix begs the question what about the `Txn`? Since it's not the transaction we care about but the potential execution time. `maxConflictingExecutionTimestamp`? A bit of a mouthful.

Reply:
Totally fine not to change it. I don't think I was confused by it you had just volunteered to change it.